src/server/tiered_storage.cc

```diff
@@ -770,4 +780,114 @@ bool TieredStorage::CanExternalizeEntry(PrimeIterator it) {
          !it->second.IsExternal() && EligibleForOffload(it->second.Size());
 }
 
+class TieredStorageV2::ShardOpManager : public tiering::OpManager {
+ public:
+  ShardOpManager(TieredStorageV2* ts, DbSlice* db_slice) : ts_{ts}, db_slice_{db_slice} {
+  }
+
+  // Find entry by key in db_slice and store external segment in place of original value
+  void SetExternal(std::string_view key, tiering::DiskSegment segment) {
+    if (auto pv = Find(key); pv) {
+      pv->SetIoPending(false);
+      pv->SetExternal(segment.offset, segment.length);  // TODO: Handle memory stats
+    }
+  }
+
+  void ClearIoPending(std::string_view key) {
+    if (auto pv = Find(key); pv)
+      pv->SetIoPending(false);
+  }
+
+  // Find entry by key and store it's up-to-date value in place of external segment
+  void SetInMemory(std::string_view key, std::string_view value) {
+    if (auto pv = Find(key); pv)
+      pv->SetString(value);
+  }
+
+  void ReportStashed(EntryId id, tiering::DiskSegment segment) override {
+    if (holds_alternative<string_view>(id)) {
+      SetExternal(get<string_view>(id), segment);
+    } else {
+      for (const auto& [sub_key, sub_segment] :
+           ts_->bins_->ReportStashed(get<tiering::SmallBins::BinId>(id), segment))
+        SetExternal(string_view{sub_key}, sub_segment);
+    }
+  }
```

Comment:
I'm not find of splitting it because it's now a two-way dependency, but it frees up the header a lot